kaldb/src/main/java/com/slack/kaldb/blobfs/s3/S3CrtBlobFs.java

```diff
@@ -0,0 +1,548 @@
+    LOG.info("mkdir {}", uri);
+    try {
+      Preconditions.checkNotNull(uri, "uri is null");
+      String path = normalizeToDirectoryPrefix(uri);
+      // Bucket root directory already exists and cannot be created
+      if (path.equals(DELIMITER)) {
+        return true;
+      }
+
+      PutObjectRequest putObjectRequest =
+          PutObjectRequest.builder().bucket(uri.getHost()).key(path).build();
+
+      PutObjectResponse putObjectResponse =
+          s3AsyncClient.putObject(putObjectRequest, AsyncRequestBody.fromBytes(new byte[0])).get();
+
+      return putObjectResponse.sdkHttpResponse().isSuccessful();
+    } catch (Throwable t) {
+      throw new IOException(t);
+    }
+  }
+
+  @Override
+  public boolean delete(URI segmentUri, boolean forceDelete) throws IOException {
+    LOG.info("Deleting uri {} force {}", segmentUri, forceDelete);
+    try {
+      if (isDirectory(segmentUri)) {
+        if (!forceDelete) {
+          Preconditions.checkState(
+              isEmptyDirectory(segmentUri),
+              "ForceDelete flag is not set and directory '%s' is not empty",
+              segmentUri);
+        }
+        String prefix = normalizeToDirectoryPrefix(segmentUri);
+        ListObjectsV2Response listObjectsV2Response;
+        ListObjectsV2Request.Builder listObjectsV2RequestBuilder =
+            ListObjectsV2Request.builder().bucket(segmentUri.getHost());
+
+        if (prefix.equals(DELIMITER)) {
+          ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.build();
+          listObjectsV2Response = s3AsyncClient.listObjectsV2(listObjectsV2Request).get();
+        } else {
+          ListObjectsV2Request listObjectsV2Request =
+              listObjectsV2RequestBuilder.prefix(prefix).build();
+          listObjectsV2Response = s3AsyncClient.listObjectsV2(listObjectsV2Request).get();
+        }
+        boolean deleteSucceeded = true;
+        for (S3Object s3Object : listObjectsV2Response.contents()) {
+          DeleteObjectRequest deleteObjectRequest =
+              DeleteObjectRequest.builder()
+                  .bucket(segmentUri.getHost())
+                  .key(s3Object.key())
+                  .build();
+
+          DeleteObjectResponse deleteObjectResponse =
+              s3AsyncClient.deleteObject(deleteObjectRequest).get();
+
+          deleteSucceeded &= deleteObjectResponse.sdkHttpResponse().isSuccessful();
+        }
+        return deleteSucceeded;
+      } else {
+        String prefix = DELIMITER + sanitizePath(segmentUri.getPath());
+        DeleteObjectRequest deleteObjectRequest =
+            DeleteObjectRequest.builder().bucket(segmentUri.getHost()).key(prefix).build();
+
+        DeleteObjectResponse deleteObjectResponse =
+            s3AsyncClient.deleteObject(deleteObjectRequest).get();
+
+        return deleteObjectResponse.sdkHttpResponse().isSuccessful();
+      }
+    } catch (NoSuchKeyException e) {
+      return false;
```

Comment:
This matches the current behavior here - https://github.com/slackhq/kaldb/blob/master/kaldb/src/main/java/com/slack/kaldb/blobfs/s3/S3BlobFs.java#L297. 

If I recall this was done to make this both reentrant, and so that you didn't need to first check if the file existed before attempting to delete. We don't have a way to guarantee that check-then-delete as atomic, so this just handles the exception for you.